src/dotnet/commands/dotnet-build/ProjectBuilder.cs

```diff
@@ -0,0 +1,88 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.DotNet.ProjectModel;
+using NuGet.Frameworks;
+
+namespace Microsoft.DotNet.Tools.Build
+{
+    internal enum CompilationResult
+    {
+        IncrementalSkip, Success, Failure
+    }
+
+    internal abstract class ProjectBuilder
+    {
+        private readonly bool _skipDependencies;
+
+        public ProjectBuilder(bool skipDependencies)
+        {
+            _skipDependencies = skipDependencies;
+        }
+
+        private Dictionary<ProjectContextIdentity, CompilationResult> _compilationResults = new Dictionary<ProjectContextIdentity, CompilationResult>();
+
+        public IEnumerable<CompilationResult> Build(IEnumerable<ProjectGraphNode> roots)
+        {
+            foreach (var projectNode in roots)
+            {
+                yield return Build(projectNode);
+            }
+        }
+
+        private CompilationResult Build(ProjectGraphNode projectNode)
+        {
+            CompilationResult result;
+            if (_compilationResults.TryGetValue(projectNode.ProjectContext.Identity, out result))
+            {
+                return result;
+            }
+            result = CompileWithDependencies(projectNode);
+
+            _compilationResults[projectNode.ProjectContext.Identity] = result;
+
+            return result;
+        }
+
+        private CompilationResult CompileWithDependencies(ProjectGraphNode projectNode)
+        {
+            var skipIncrementalCheck = false;
+            if (!_skipDependencies)
+            {
+                foreach (var dependency in projectNode.Dependencies)
+                {
+                    var context = dependency.ProjectContext;
+                    if (!context.ProjectFile.Files.SourceFiles.Any())
+                    {
+                        continue;
+                    }
+                    var result = Build(dependency);
+                    switch (result)
+                    {
+                        case CompilationResult.IncrementalSkip:
+                            break;
+                        case CompilationResult.Success:
+                            skipIncrementalCheck = true;
+                            break;
+                        case CompilationResult.Failure:
+                            return CompilationResult.Failure;
+                        default:
+                            throw new NotSupportedException();
+                    }
+                }
+            }
+            if (skipIncrementalCheck || NeedsRebuilding(projectNode))
```

Comment:
Yes, if inputs are written outside of NeedsRebuild (like besides SDK stamping) then there's no stale cache with the `skipIncrementalCheck` short circuit. 

The downside to separating caching inputs from the actual incremental check is that it scatters the responsibility around. For example, responding to the --no-incremental is harder, since more parts of the code would need to branch on it, not just NeedsRebuild.

---
In reply to: [60173794](https://github.com/dotnet/cli/pull/2574#discussion_r60173794) [](ancestors = 60173794)